### ISSUE TITLE
Correctness difference between HF GPT2Block and TransformerLayer

### ISSUE THREAD

**Issue by Teng-xu** [NVIDIA/TransformerEngine#413]
I am attempting to utilize the TransformerLayer for executing the HF GPT2 model. I've developed a translation function to convert the HF model parameters into TransformerLayer parameters. However, I'm observing disparities between the output of HF's GPT2Block and that of the TransformerLayer. This difference is more obvious on A100 machines compared to V100 machines. When using the bf16 detype, the differences in output are further amplified, larger than the typical differences we would get from variations in precision. 
I am trying to indentify the potential factors causing these discrepancies and exploring possible remedies. Your insights on identifying the underlying reasons and potential solutions would be greatly appreciated.

Here is the repro script:
```
import argparse

import numpy as np
import torch
from transformers import (
    GPT2Config,
    GPT2LMHeadModel,
    GPT2Tokenizer,
)
from transformers.models.gpt2.modeling_gpt2 import GPT2Block
from transformer_engine.pytorch import TransformerLayer as TETransformerLayer
from transformer_engine.pytorch.utils import init_method_normal


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--atol", type=float, default=1e-3)
    parser.add_argument("--bf16", type=int, default=0)
    args, _ = parser.parse_known_args()
    return args


def test_gpt2(atol, bf16):
    torch.manual_seed(0)

    gpt2_config = GPT2Config(n_layer=1, use_cache=False)

    hf_layer = GPT2Block(gpt2_config)
    hf = hf_layer.state_dict()

    te_dict = translate(hf, gpt2_config.n_head)
    # print('hf_layer: ', hf_layer)

    hf_layer.to(0)
    hf_layer.eval()
    input_ids = torch.rand(1, 6, 768).to(0)

    hf_output = hf_layer(hidden_states=input_ids, attention_mask=None)[0]

    te_layer = TETransformerLayer(
        hidden_size=gpt2_config.n_embd,
        ffn_hidden_size=gpt2_config.n_embd*4,
        num_attention_heads=gpt2_config.n_head,
        layernorm_epsilon=gpt2_config.layer_norm_epsilon,
        hidden_dropout=gpt2_config.resid_pdrop,
        attention_dropout=gpt2_config.attn_pdrop,
        init_method=init_method_normal(gpt2_config.initializer_range),
        output_layer_init_method=init_method_normal(gpt2_config.initializer_range),
        layer_number=1,
        self_attn_mask_type="causal",
        tp_group=None,
        tp_size=1,
        params_dtype=torch.bfloat16 if bf16 > 0 else torch.float32,
        get_rng_state_tracker=None,
        fuse_wgrad_accumulation=False,
        seq_length=None,
        micro_batch_size=None,
        sequence_parallel=True,
        layer_type="encoder",
        set_parallel_mode=True,
        fuse_qkv_params=True,
        qkv_weight_interleaved=True,
    )

    te_layer.to(0)
    te_layer.load_state_dict(te_dict, strict=False)
    if bf16>0:
        input_ids = input_ids.to(dtype=torch.bfloat16).transpose(0, 1).contiguous()
    else:
        input_ids = input_ids.transpose(0, 1).contiguous()

    te_layer.eval()
    te_output = te_layer(hidden_states=input_ids, attention_mask=None)

    hf_output = hf_output.detach().cpu()
    te_output = te_output.transpose(0, 1).detach().cpu()
    np.testing.assert_allclose(hf_output.float(), te_output.float(), atol=atol)


def translate(state_dict, num_attention_heads):
    translated_state_dict = {}

    for name, param in state_dict.items():
        tokens = name.split(".")
        param_type = tokens[-1]

        block = tokens[0]

        if block == "mlp":
            layer = tokens[1]
            dense_idx = "1" if layer == "c_fc" else "2"

            translated_name = (
                "layernorm_mlp.fc"
                + dense_idx
                + "_"
                + param_type
            )
            if param_type == "weight":
                translated_param = param.t()
            else:
                translated_param = param

            translated_state_dict[translated_name] = translated_param

        elif block == "attn":
            layer = tokens[1]
            if layer == "c_proj":
                translated_name = (
                    "self_attention.proj." + param_type
                )
                if param_type == "weight":
                    translated_param = param.t()
                else:
                    translated_param = param

                translated_state_dict[translated_name] = translated_param
            elif layer == "c_attn":
                translated_name = (
                    "self_attention.layernorm_qkv." + param_type
                )
                if param_type in ["weight", "bias"]:
                    if param_type == "weight":
                        param = param.t()
                    attention_size = param.size(0) // 3
                    attention_head_size = attention_size // num_attention_heads
                    new_param = torch.clone(param)
                    translated_param = torch.clone(param)

                    q = new_param.narrow(0, 0, attention_size)
                    k = new_param.narrow(0, attention_size, attention_size)
                    v = new_param.narrow(0, 2 * attention_size, attention_size)

                    for i in range(num_attention_heads):
                        start = i * attention_head_size * 3
                        qkv_start = i * attention_head_size
                        translated_param.narrow(0, start, attention_head_size).copy_(
                            q.narrow(0, qkv_start, attention_head_size)
                        )
                        translated_param.narrow(
                            0, start + attention_head_size, attention_head_size
                        ).copy_(k.narrow(0, qkv_start, attention_head_size))
                        translated_param.narrow(
                            0, start + attention_head_size * 2, attention_head_size
                        ).copy_(v.narrow(0, qkv_start, attention_head_size))
                translated_state_dict[translated_name] = translated_param
        elif block in ["ln_1", "ln_2"]:
            att_vs_out = "self_attention.layernorm_qkv" if block == "ln_1" else "layernorm_mlp"
            translated_name = (
                att_vs_out
                + ".layer_norm_"
                + param_type
            )
            translated_state_dict[translated_name] = param

    return translated_state_dict


def main():
    args = parse_args()

    test_gpt2(args.atol, args.bf16)


if __name__ == "__main__":
    main()

```

**Comment by ptrendx** [comment]
Hi @Teng-xu. I believe in your case you should set `qkv_weight_interleaved` to `False` to match HF implementation of GPT2.

Here is an example comparing a single Transformer layer (without dropout to compare the results):
```python
import transformers
import transformer_engine.pytorch as te
import torch

config = transformers.models.gpt2.configuration_gpt2.GPT2Config()

config.attn_pdrop = 0
config.embd_pdrop = 0
config.resid_pdrop = 0
config.n_positions = 2048

hf_layer = transformers.models.gpt2.modeling_gpt2.GPT2Block(config)
hf_layer.cuda()

te_layer = te.TransformerLayer(config.hidden_size,
                               4 * config.hidden_size,
                               config.n_head,
                               layernorm_epsilon=1e-05,
                               hidden_dropout=0,
                               attention_dropout=0,
                               fuse_qkv_params=True,
                               qkv_weight_interleaved=False
                              )

def reset_params(hf_layer, te_layer):
    te_layer.self_attention.layernorm_qkv.layer_norm_weight.data[:] = hf_layer.ln_1.weight.data[:]
    te_layer.self_attention.layernorm_qkv.layer_norm_bias.data[:] = hf_layer.ln_1.bias.data[:]
    te_layer.self_attention.layernorm_qkv.weight.data[:] = hf_layer.attn.c_attn.weight.data[:].transpose(1,0)
    te_layer.self_attention.layernorm_qkv.bias.data[:] = hf_layer.attn.c_attn.bias.data[:]
    te_layer.self_attention.proj.weight.data[:] = hf_layer.attn.c_proj.weight.data[:].transpose(1,0)
    te_layer.self_attention.proj.bias.data[:] = hf_layer.attn.c_proj.bias.data[:]
    te_layer.layernorm_mlp.layer_norm_weight.data[:] = hf_layer.ln_2.weight.data[:]
    te_layer.layernorm_mlp.layer_norm_bias.data[:] = hf_layer.ln_2.bias.data[:]
    te_layer.layernorm_mlp.fc1_weight.data[:] = hf_layer.mlp.c_fc.weight.data[:].transpose(1,0)
    te_layer.layernorm_mlp.fc1_bias.data[:] = hf_layer.mlp.c_fc.bias.data[:]
    te_layer.layernorm_mlp.fc2_weight.data[:] = hf_layer.mlp.c_proj.weight.data[:].transpose(1,0)
    te_layer.layernorm_mlp.fc2_bias.data[:] = hf_layer.mlp.c_proj.bias.data[:]

reset_params(hf_layer, te_layer)
hf_layer.bfloat16()
te_layer.bfloat16()

x = torch.rand((4, 2048, config.hidden_size)).cuda().bfloat16()
xt = x.transpose(0,1).contiguous()

y = hf_layer(x)
y_te = te_layer(xt).transpose(0,1).contiguous()

print(torch.max(torch.abs(y[0] - y_te)))
```

On my machine this gives:
```
tensor(0.0156, device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>)
```

**Comment by Teng-xu** [comment]
Thanks @ptrendx, but I think the issue here is not about `qkv_weight_interleaved`, I think my translation function has taken care of the qkv_weight_interleaved weight transformation, and with my script I am also able to get `tensor(0.0131)` with `torch.max(torch.abs(hf_output - te_output)))`, however, my concern is whether 0.0131 or 0.0156 are still excessively large values in terms of convergence? Because I am able to get `tensor(5.9605e-07)` on V100 machines and with fp32, which is significantly smaller.

**Comment by ptrendx** [comment]
Hi @Teng-xu, bfloat16 data format has significantly less bits of mantissa than FP32, and so the relative difference between values will be much bigger. Intuitively, since B16 is E8M7 format, the errors would be on the order of `2**(-7) ~ 1e-2`. In practice you can see this by comparing the outputs of the BF16 model with the FP32 model (just remember, even for the FP32 model, to cast the model weights and the input first to bf16 in order for the models to work on the same data). Here is the modified version of the script I gave in the previous comment that does this (it requires applying PR #414 for TE that fixes a bug with models that are cast multiple times):
```python
import transformers
import transformer_engine.pytorch as te
import torch

config = transformers.models.gpt2.configuration_gpt2.GPT2Config()

config.attn_pdrop = 0
config.embd_pdrop = 0
config.resid_pdrop = 0
config.n_positions = 2048

hf_layer = transformers.models.gpt2.modeling_gpt2.GPT2Block(config)
hf_layer.cuda()

te_layer = te.TransformerLayer(config.hidden_size,
                               4 * config.hidden_size,
                               config.n_head,
                               layernorm_epsilon=1e-05,
                               hidden_dropout=0,
                               attention_dropout=0,
                               fuse_qkv_params=True,
                               qkv_weight_interleaved=False
                              )

def reset_params(hf_layer, te_layer):
    te_layer.self_attention.layernorm_qkv.layer_norm_weight.data[:] = hf_layer.ln_1.weight.data[:]
    te_layer.self_attention.layernorm_qkv.layer_norm_bias.data[:] = hf_layer.ln_1.bias.data[:]
    te_layer.self_attention.layernorm_qkv.weight.data[:] = hf_layer.attn.c_attn.weight.data[:].transpose(1,0)
    te_layer.self_attention.layernorm_qkv.bias.data[:] = hf_layer.attn.c_attn.bias.data[:]
    te_layer.self_attention.proj.weight.data[:] = hf_layer.attn.c_proj.weight.data[:].transpose(1,0)
    te_layer.self_attention.proj.bias.data[:] = hf_layer.attn.c_proj.bias.data[:]
    te_layer.layernorm_mlp.layer_norm_weight.data[:] = hf_layer.ln_2.weight.data[:]
    te_layer.layernorm_mlp.layer_norm_bias.data[:] = hf_layer.ln_2.bias.data[:]
    te_layer.layernorm_mlp.fc1_weight.data[:] = hf_layer.mlp.c_fc.weight.data[:].transpose(1,0)
    te_layer.layernorm_mlp.fc1_bias.data[:] = hf_layer.mlp.c_fc.bias.data[:]
    te_layer.layernorm_mlp.fc2_weight.data[:] = hf_layer.mlp.c_proj.weight.data[:].transpose(1,0)
    te_layer.layernorm_mlp.fc2_bias.data[:] = hf_layer.mlp.c_proj.bias.data[:]

reset_params(hf_layer, te_layer)
hf_layer.bfloat16()
te_layer.bfloat16()

x = torch.rand((4, 2048, config.hidden_size)).cuda().bfloat16()
xt = x.transpose(0,1).contiguous()

y = hf_layer(x)
y_te = te_layer(xt).transpose(0,1).contiguous()

#reference
hf_layer.float()
te_layer.float()
x_float = x.float()
xt_float = xt.float()
y_hf_ref = hf_layer(x_float)
y_te_ref = te_layer(xt_float).transpose(0,1).contiguous()

print(f"HF vs TE fp32: {torch.max(torch.abs(y_hf_ref[0] - y_te_ref))}")

print(f"HF vs TE BF16: {torch.max(torch.abs(y[0] - y_te))}")
print(f"HF FP32 vs HF BF16: {torch.max(torch.abs(y_hf_ref[0] - y[0]))}")
print(f"HF FP32 vs TE BF16: {torch.max(torch.abs(y_hf_ref[0] - y_te))}")
print(f"TE FP32 vs TE BF16: {torch.max(torch.abs(y_te_ref - y_te))}")
```
This gives me outputs like this:
```
HF vs TE fp32: 0.0004467964172363281
HF vs TE BF16: 0.015625
HF FP32 vs HF BF16: 0.015078544616699219
HF FP32 vs TE BF16: 0.016489267349243164
TE FP32 vs TE BF16: 0.016467571258544922
```
where you can see that both HF and TE BF16 results are similarly different from the FP32 reference.

**Comment by Teng-xu** [comment]
Yeah that make sense, I appreciate your response! @ptrendx
